src/io/category_encoding_provider.cpp

```diff
@@ -0,0 +1,1330 @@
+  std::discrete_distribution<int> fold_distribution(fold_probs.begin(), fold_probs.end());
+  std::vector<std::vector<std::unique_ptr<CSC_RowIterator>>> thread_csc_iters(num_threads_);
+  // copy csc row iterators for each thread
+  #pragma omp parallel for schedule(static) num_threads(num_threads_)
+  for (int thread_id = 0; thread_id < num_threads_; ++thread_id) {
+    for (size_t i = 0; i < csc_iters.size(); ++i) {
+      thread_csc_iters[thread_id].emplace_back(new CSC_RowIterator(*csc_iters[i].get()));
+    }
+  }
+  Threading::For<int32_t>(0, nrow, 1024,
+    [this, &thread_csc_iters, &get_label_fun, ncol, &fold_distribution, &mt_generators]
+    (int thread_id, int32_t start, int32_t end) {
+    std::vector<double> oneline_features(ncol, 0.0f);
+    for (int32_t row_idx = start; row_idx < end; ++row_idx) {
+      for (int32_t col_idx = 0; col_idx < ncol; ++col_idx) {
+        oneline_features[col_idx] = thread_csc_iters[thread_id][col_idx]->Get(row_idx);
+      }
+      const int fold_id = fold_distribution(mt_generators[thread_id]);
+      training_data_fold_id_[row_idx] = fold_id;
+      const double label = static_cast<double>(get_label_fun(row_idx));
+      ProcessOneLine(oneline_features, label, row_idx, thread_id, fold_id);
+    }
+  });
+  FinishProcess(1, config);
+}
+
+void CategoryEncodingProvider::SetConfig(const Config* config) {
+  config_ = *config;
+  num_threads_ = config_.num_threads > 0 ? config_.num_threads : OMP_NUM_THREADS();
+  keep_raw_cat_method_ = false;
+  const std::string target_encoding_string = std::string("target");
+  category_encoders_.clear();
+  if (config_.category_encoders.size() > 0) {
+    for (auto token : Common::Split(config_.category_encoders.c_str(), ',')) {
+      if (Common::StartsWith(token, "target")) {
+        if (token.size() == target_encoding_string.size()) {
+          category_encoders_.emplace_back(new CategoryEncodingProvider::TargetEncoderLabelMean());
+        } else {
+          double prior = 0.0f;
+          if (token[target_encoding_string.size()] != ':' ||
+              !Common::AtofAndCheck(token.c_str() + target_encoding_string.size() + 1, &prior)) {
+            Log::Fatal("Target encoding prior of cat_converter specification %s is not a valid float value.", token.c_str());
+          }
+          category_encoders_.emplace_back(new CategoryEncodingProvider::TargetEncoder(prior));
+        }
+      } else if (token == std::string("count")) {
+        category_encoders_.emplace_back(new CategoryEncodingProvider::CountEncoder());
+      } else if (token == std::string("raw")) {
+        keep_raw_cat_method_ = true;
+      } else {
+        Log::Fatal("Unknown category_encoders specification %s.", token.c_str());
+      }
+    }
+  }
+
+  prior_weight_ = config_.prior_weight;
+  tmp_parser_ = nullptr;
+}
+
+std::string CategoryEncodingProvider::DumpToJSON() const {
```

Comment:
Same as converter, Looks like unified single pair serialization/deserialization would reduce lots of complexity.